controller/service.go

```diff
@@ -104,6 +111,27 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 			c.clearServiceState(key, svc)
 			lbIPs = []net.IP{}
 		}
+
+		// If svc currently has 1 ip and policy PreferDualStack, try assigning ip from the missing family and same pool
+		if len(lbIPs) == 1 && familyPolicy == v1.IPFamilyPolicyPreferDualStack {
+			level.Info(l).Log("event", "tryAssignAdditionalIP", "msg", "familyPolicy is PreferDualStack, trying to assign additional ip")
+			existingFamily := ipfamily.ForAddress(lbIPs[0])
+			var additionalFamily ipfamily.Family
+			if existingFamily == ipfamily.IPv4 {
+				additionalFamily = ipfamily.IPv6
+			} else {
+				additionalFamily = ipfamily.IPv4
+			}
+			desiredPool := valueForAnnotation(svc.Annotations, AnnotationIPAllocateFromPool, DeprecatedAnnotationIPAllocateFromPool)
```

Comment:
I'm not followed why we have a pool here? This is for the case when an ip has been assigned before, but then the pool suddenly has dual-stack. The existing pool should be in the annotation, but it was assigned in another iteration, so its value is not defined.

Reply:
ah, I misunderstood it because `desiredPool` in this file is used when we want to understand what is the user's desired pool via `AnnotationAddressPool` and here you used `AnnotationIPAllocateFromPool`. In any case we shouldn't rely on an annotation to understand what pool is used for the service, it should be possible to fetch from the inner structs (the allocator or something related)

Reply:
Thank you. That was my bad, I picked the name `desiredPool` without looking further :smile: I'll change the name then.
Anw, I'm not sure if the allocator or something else (other than the svc) stores the info of which pool was used in assigning :-? I'll need to dig a bit, to my impression only the assigned ip is stored.

Reply:
Hi @oribon Thanks to your comment I was able to fix this, now the pool is taken from `c.ips.Pool(key)` instead of svc annotation. Could you please give me another round of review to see if I made it correctly :smile: ?